compiler-rt: Provide __cpu_indicator_init, __cpu_model and __cpu_features2

lacc97 opened the pull request:
Fixes https://github.com/ziglang/zig/issues/18074, original PR was https://github.com/ziglang/zig/pull/18193.

This PR adds support for X86/X86_64 CPU model detection and function multiversioning in C/C++ via the target and target_clones attributes (compatible with LLVM's, which in turn is compatible with libgcc's). It adds `__cpu_indicator_init`, `__cpu_model` and `__cpu_features2` symbols to compiler-rt for x86/x86_64 which reuse the CPU detection logic under std/zig/system/x86.zig and transform the detected features into the values expected by LLVM-generated code. These values are kept in sync with LLVM using a new tool (tools/update_x86_cpu_model_enums.zig) which parses llvm/include/llvm/TargetParser/X86TargetParser.def in LLVM to generate the corresponding Zig enums.

This PR also adds a standalone test that the symbols are emitted when the `target` attribute is used in C.

While this functionality could in principle contribute to solving ~~[issue #4591](https://github.com/ziglang/zig/issues/4591)~~ (edit: wrong issue, its #1018), I think it would be best to use a Zig-specific solution for richer CPU detection capabilities, leveraging std.Target.* instead of relying only on the feature set supported by LLVM/libgcc. The changes in this PR would therefore only be useful for C/C++ compatibility with Clang and gcc.

lacc97 committed:
compiler-rt: Provide LLVM CPU detection symbols for x86_64

These match what LLVM expects and generates when building multiversioned
functions (e.g. functions tagged with attributes target or
target_clones). The actual CPU detection reuses the logic under
std/zig/system/x86.zig, and transforms it into the values that
LLVM-generated code expects. These values are auto-generated using a
new tool (tools/update_x86_cpu_model_enums.zig) which parses the
relevant file in the LLVM codebase and generates Zig enums out of it.

lacc97 committed:
update_x86_cpu_model_enums: Fix eating first character in enum name

lacc97 committed:
compiler_rt: Cleanup compile error output in x86_cpu_model

lacc97 committed:
compiler_rt: Update x86 cpu enums for LLVM 18

lacc97 committed:
compiler_rt: Support x86_64{,_v2,_v3,_v4} feature detection for LLVM

lacc97 committed:
compiler_rt: Documentation

lacc97 committed:
compiler_rt: Add Intel Clearwaterforest and clarify AMD detection

lacc97 committed:
compiler_rt: Update autogenerated file

daurnimator commented:
> I could have sworn there was an issue about function multiversioning/selection based on runtime-detected CPU in Zig itself but I can't find it now

https://github.com/ziglang/zig/issues/1018 ?

lacc97 commented:
> > I could have sworn there was an issue about function multiversioning/selection based on runtime-detected CPU in Zig itself but I can't find it now
> 
> https://github.com/ziglang/zig/issues/1018 ?
> 
> 

Yep that was it. Thanks.

theseyan commented:
Missing support for these symbols is a blocker for compiling `libmdbx` (a fork of LMDB) with Zig:
```bash
error: ld.lld: undefined symbol: __cpu_indicator_init
    note: referenced by gc-get.c:584 (src/gc-get.c:584)
    note:               mdbx-static.o:(scan4seq_resolver) in archive [...]/libmdbx.a
error: ld.lld: undefined symbol: __cpu_model
    note: referenced by gc-get.c:587 (src/gc-get.c:587)
    note:               mdbx-static.o:(scan4seq_resolver) in archive [...]/libmdbx.a
```
I really appreciate this PR and hope it gets merged soon. 🫡

slyshykO commented:
As a temporary solution, it is possible to use [cpu_model](https://github.com/slyshykO/cpu_model)

lacc97 committed:
compiler_rt: Update x86_cpu_model.zig to latest zig master

- Use pointers in export builtin
- Update names for Type enum tag

lacc97 committed:
compiler_rt: Update x86 cpu model enums to LLVM 19

lacc97 commented:
Rebased to latest master.

lacc97 commented:
Rebased again . I'm not entirely happy with `getAmdTypeAndSubtype()` and `getIntelTypeAndSubtype()` as it seems to duplicate similar logic as in `lib/std/zig/system/x86.zig`. It's not clear to me how to resolve this without bringing in the generated enum types inside the standard library, which don't really fit there.

lacc97 committed:
tools: Fix update_x86_cpu_model_enums logic for calculating enum value

The new logic is closer to how it's done in C where any unspecified
enum simply increments the previous value.

alexrp commented:
> It's not clear to me how to resolve this without bringing in the generated enum types inside the standard library, which don't really fit there.

Is it actually problematic if the standard library `@import`s a file from compiler-rt or vice versa? If it just contains data, as seems to be the case, that seems to me like it'd be harmless?

lacc97 committed:
compiler_rt: Move generated x86 cpu enums under std

This lets us reuse the existing cpu detection in std.zig.system.x86.
Hence, lib/compiler_rt/x86_cpu_model.zig simply translates the
detected Target.Cpu.Feature.Set into the features set expected by LLVM.

lacc97 commented:
> > It's not clear to me how to resolve this without bringing in the generated enum types inside the standard library, which don't really fit there.
> 
> Is it actually problematic if the standard library `@import`s a file from compiler-rt or vice versa? If it just contains data, as seems to be the case, that seems to me like it'd be harmless?

Made a new commit with regards to this. I moved the enum definition file into lib/std/zig/system and modified some of lib/std/zig/system/x86.zig functions to return the Vendor, Type and Subtype enums as part of detection in addition to the pointer to Target.Cpu.Model. Some of these decls are also exposed so that the compiler-rt code can use them. Now, the only logic in the compiler-rt file simply converts between Target.Cpu.Feature.Set and the actual values expected by LLVM for the `target`/`target_clones` attributes to work.

andrewrk commented:
@alexrp are you interested in shepherding this one?

alexrp commented on lib/compiler_rt/x86_cpu_model.zig:
```diff
@@ -0,0 +1,165 @@
+// TODO: Do we need constructor attribute? As far as I can tell the generated LLVM detection code explicitly
+//       calls __cpu_indicator_init before actually using __cpu_model or __cpu_features2 (happens in
+//       clang/lib/CodeGen/CGBuiltin.cpp)
+fn init() callconv(.C) c_int {
+    if (cpu.vendor != .unknown) return 0;
```
Since this can be called from multiple threads, this check should be replaced with an atomic flag using unordered loads/stores.

lacc97 replied:
Should it be acquire-release instead to ensure ordering with respect to the other fields?

andrewrk commented:
I do want a final review because it looks very messy

alexrp commented on lib/compiler_rt/x86_cpu_model.zig:
```diff
@@ -0,0 +1,165 @@
+};
+
+comptime {
+    // Sanity checking that the feature sets in std.Target.x86 and the autogenerated
+    // ones in @This().Feature are compatible.
```
I don't love the idea of doing this here due to compile times. Since we will be running the new script during [the LLVM upgrade process](https://github.com/ziglang/zig/wiki/LLVM-Upgrade-Process) much like `update_cpu_features.zig`, I think it would make more sense to just do these sanity checks in the script and require that it's built with a compiler that has an up-to-date `std.Target.x86`.

alexrp replied:
You might even consider merging the new script into `update_cpu_features.zig` since they should both be run together anyway.

alexrp commented on lib/compiler_rt/x86_cpu_model.zig:
```diff
@@ -0,0 +1,165 @@
+var cpu: Model = .{};
+var cpu_extra_features: [feature_set_size - 1]u32 = [_]u32{0} ** (feature_set_size - 1);
+
+// TODO: Do we need constructor attribute? As far as I can tell the generated LLVM detection code explicitly
+//       calls __cpu_indicator_init before actually using __cpu_model or __cpu_features2 (happens in
+//       clang/lib/CodeGen/CGBuiltin.cpp)
```
Judging by the relevant Clang code, this doesn't seem necessary.

```suggestion
```

alexrp commented on test/standalone/x86_cpu_model/build.zig:
```diff
@@ -0,0 +1,57 @@
+    if (!arch.isX86()) return;
+
+    const abi = target.result.abi;
+    if (target.result.ofmt != .elf or !(abi.isMusl() or abi.isGnu())) return;
```
Why limit to these targets?

alexrp replied:
Oh, I see. On macOS:

```
/Users/runner/work/zig/zig/build/../lib/std/Build/Step/CheckObject.zig:506:17: 0x10a7671e1 in checkInDynamicSymtab (build)
        else => @panic("Unsupported target platform"),
```

But really, since the test is now actually testing multiversioning at runtime, is there any value in doing these symbol checks anyway?

alexrp commented on test/standalone/x86_cpu_model/build.zig:
```diff
@@ -0,0 +1,57 @@
+    const abi = target.result.abi;
+    if (target.result.ofmt != .elf or !(abi.isMusl() or abi.isGnu())) return;
+
+    const exe = b.addExecutable(.{
+        .name = "main",
+        .optimize = optimize,
+        .target = target,
+    });
```
I would suggest adding an extra step here that actually runs the executable. The test program itself should return the value from `foo_impl` as the exit code from `main`. Then you can make the build script expect the appropriate exit code depending on the CPU features of the build host.

lacc97 committed:
compiler_rt: Remove TODO

lacc97 committed:
compiler_rt: Use atomics for loading cpu model

lacc97 committed:
compiler_rt: Check x86_cpu_model standalone test binary

Checks the exit code which depends on the running
cpu feature set.

lacc97 committed:
tools: Remove update_x86_cpu_model

Move the functionality and code to update_cpu_features.

lacc97 committed:
tools: Reorganize x86 enum processing in update_cpu_features

lacc97 committed:
Move x86 cpu feature compat check to tools/update_cpu_features

Instead of checking this at comptime, they are checked
at the point of generation in tools/update_cpu_features.zig.

alexrp commented on test/standalone/x86_cpu_model/build.zig:
```diff
@@ -34,6 +33,17 @@ pub fn build(b: *std.Build) void {
     test_step.dependOn(&check_exe.step);
 
+    const run_exe = b.addRunArtifact(exe);
+    if (std.Target.x86.featureSetHas(cpu.features, .avx512vnni)) {
+        run_exe.expectExitCode(3);
+    } else if (std.Target.x86.featureSetHas(cpu.features, .avx2)) {
+        run_exe.expectExitCode(2);
+    } else {
+        run_exe.expectExitCode(1);
+    }
```
This'll work for most cases but it isn't quite correct: Multiversioning will dynamically detect the features of the system the program is run on. The current approach breaks if I invoke this test manually with `-Dcpu=...`. So really, you should be checking `b.graph.host` features here.

andrewrk commented:
This pull request is not ready for review because:
* It has conflicts that must be resolved via rebasing against latest origin/master.
* It is not passing the CI tests.
* No response to core team feedback for 30+ days.
